### ISSUE TITLE
expat-config-version.cmake contains bitness-specific information

### ISSUE THREAD

**Issue by ryandesign** [libexpat/libexpat#875]
The installed file $PREFIX/lib/cmake/expat-2.6.2/expat-config-version.cmake contains bitness-specific information. On my 64-bit build of expat it contains:

```cmake
# if the installed or the using project don't have CMAKE_SIZEOF_VOID_P set, ignore it:
if("${CMAKE_SIZEOF_VOID_P}" STREQUAL "" OR "8" STREQUAL "")
  return()
endif()

# check that the installed version has the same 32/64bit-ness as the one which is currently searching:
if(NOT CMAKE_SIZEOF_VOID_P STREQUAL "8")
  math(EXPR installedBits "8 * 8")
  set(PACKAGE_VERSION "${PACKAGE_VERSION} (${installedBits}bit)")
  set(PACKAGE_VERSION_UNSUITABLE TRUE)
endif()
```

On a 32-bit build it contains:

```cmake
# if the installed or the using project don't have CMAKE_SIZEOF_VOID_P set, ignore it:
if("${CMAKE_SIZEOF_VOID_P}" STREQUAL "" OR "4" STREQUAL "")
  return()
endif()

# check that the installed version has the same 32/64bit-ness as the one which is currently searching:
if(NOT CMAKE_SIZEOF_VOID_P STREQUAL "4")
  math(EXPR installedBits "4 * 8")
  set(PACKAGE_VERSION "${PACKAGE_VERSION} (${installedBits}bit)")
  set(PACKAGE_VERSION_UNSUITABLE TRUE)
endif()
```

This is a problem for macOS on which it is common to want a [universal binary](https://en.wikipedia.org/wiki/Universal_binary) which is compiled for multiple architectures, such as i386-and-x86_64 or x86_64-and-arm64.

Can you please remove all bitness- and architecture-specific information from expat-config-version.cmake so that that single file can be used regardless of the bitness or architecture for which one is building something that uses expat?

**Comment by hartwork** [comment]
Hi @ryandesign,

good to see you! I believe this is about https://github.com/macports/macports-ports/blob/master/textproc/expat/Portfile which seems to use the GNU Autotools build system of libexpat. The Autotools build systems installs CMake files that resemble what CMake itself would produce down to the last character, and we have [CI that specifically ensures these files remain in sync](https://github.com/libexpat/libexpat/blob/master/.github/workflows/autotools-cmake.yml). If we change it for the Autotools side of things only, it would have the same issues for all distros using the CMake build systems unfixed and also we'd need to find ways to teach this new difference in installed CMake files and that difference only, to not miss out on detecting future going-out-of-sync regressions. I understand this is not ideal, but I only see pest or cholera here, and then we can just stay with pest, unless someone has an idea to bump things to flue level at least, if that who illness analogy makes sense. This is not totally unlike #501 btw. What do you think?

Best, Sebastian

**Comment by ryandesign** [comment]
I have no objection to keeping the files in sync between the cmake and autotools build systems. So change both the autotools and cmake build systems to remove any bitness-related information from the installed files. If you need to know bitness, you can discover it at compile time using C preprocessor macros. `__LP64__` is defined on 64-bit systems and undefined on 32-bit systems.

**Comment by ryandesign** [comment]
> This is not totally unlike #501 btw. What do you think?

I don't see any similarity between the two issues.

**Comment by hartwork** [comment]
> I have no objection to keeping the files in sync between the cmake and autotools build systems. So change both the autotools and cmake build systems to remove any bitness-related information from the installed files.

@ryandesign the thing is that CMake generates the files like that, it's CMake nit libexpat controlling these details. And that's the details that Automake just mimcs here. How do we change the CMake-controlled side of this?

> > This is not totally unlike #501 btw. What do you think?
> 
> I don't see any similarity between the two issues.

The shared part is CMake files about to be installed, that it's not moving before a good idea, and that touching it will break CI, unless second good idea.

Are there any ideas?

**Comment by hartwork** [comment]
@ryandesign PS: This problem is likely bigger than just libexpat:

```console
# grep -l "CMAKE_SIZEOF_VOID_P.*8" /usr/lib64/cmake/*/*-config-version.cmake
/usr/lib64/cmake/capstone/capstone-config-version.cmake
/usr/lib64/cmake/check/check-config-version.cmake
/usr/lib64/cmake/cmocka/cmocka-config-version.cmake
/usr/lib64/cmake/dbusmenu-qt5/dbusmenu-qt5-config-version.cmake
/usr/lib64/cmake/expat-2.6.2/expat-config-version.cmake
/usr/lib64/cmake/fmt/fmt-config-version.cmake
/usr/lib64/cmake/glslang/glslang-config-version.cmake
/usr/lib64/cmake/kissfft/kissfft-config-version.cmake
/usr/lib64/cmake/lapack-3.12.0/lapack-config-version.cmake
/usr/lib64/cmake/libssh/libssh-config-version.cmake
/usr/lib64/cmake/protobuf/protobuf-config-version.cmake
/usr/lib64/cmake/pugixml/pugixml-config-version.cmake
/usr/lib64/cmake/rnp/rnp-config-version.cmake
/usr/lib64/cmake/rpm/rpm-config-version.cmake
/usr/lib64/cmake/simdjson/simdjson-config-version.cmake
/usr/lib64/cmake/tinyxml2/tinyxml2-config-version.cmake
```

**Comment by hartwork** [comment]
@ryandesign PPS: Or compiling twice is just the answer…

```console
# grep -l "CMAKE_SIZEOF_VOID_P.*4" /usr/lib/cmake/*/*-config-version.cmake
/usr/lib/cmake/expat-2.6.2/expat-config-version.cmake
```

**Comment by ryandesign** [comment]
For universal builds, the MacPorts expat port is already configured to compile twice, once for each architecture. A post-processing step then attempts to merge all of the resulting files. There is no way to merge differing cmake files. At present, MacPorts erroneously attempts to merge the files using C preprocessor syntax, resulting in a syntactically invalid file that does not work; a user [reporting this to MacPorts](https://trac.macports.org/ticket/70228) is what prompted me to file this bug report with you. The fix for the erroneous MacPorts behaviour would be to make MacPorts emit an error if cmake files from different architectures of a universal build were found to differ. This would prevent the invalid file from being installed but it would also prevent the user from being able to install expat universally when that is what they desire to do.

The examples you show above are from a system that has separate 32-bit (lib) and 64-bit (lib64) directories. The macOS filesystem is not arranged in that manner. There is only ever a single lib directory and it can contain libraries that contain any number of different architectures.

I do not know how to fix the problem in your build system or in cmake, wherever the problem may be; I can only report to you that the problem exists and what its consequences are.

**Comment by ryandesign** [comment]
> This problem is likely bigger than just libexpat

I agree there are a lot of files on my system from different software containing this same code. I guess I'll have to take it up with the developers of cmake.

**Comment by hartwork** [comment]
> The examples you show above are from a system that has separate 32-bit (lib) and 64-bit (lib64) directories. The macOS filesystem is not arranged in that manner. There is only ever a single lib directory and it can contain libraries that contain any number of different architectures.

@ryandesign that's key, I understand. My sample was from a multilib Linux box, correct.

> > This problem is likely bigger than just libexpat
> 
> I agree there are a lot of files on my system from different software containing this same code. I guess I'll have to take it up with the developers of cmake.

Thank you! :+1:
